src/cca/scripts/uproxy_server.ts

```diff
@@ -56,25 +37,79 @@ export class UproxyServer implements Server {
   }
 }
 
+// Name by which servers are saved to storage.
+const SERVERS_STORAGE_KEY = 'saved-servers';
+
+// Type of the object placed, in serialised form, in storage.
+interface SavedServers {
+  servers: { [id: string]: SavedServer }
+}
+
+// A server as saved to storage.
+interface SavedServer {
+  cloudTokens?: cloud_social_provider.Invite;
+}
+
+// Maintains a persisted set of servers and liases with the core.
 export class UproxyServerRepository implements ServerRepository {
-  constructor(private core: CoreConnector, private vpnDevice: VpnDevice) {
-    this.core.login({
-      network: 'Cloud',
-      loginType: uproxy_core_api.LoginType.INITIAL,
-    });
+  // core must already be logged into social networks.
+  constructor(
+    private storage: Storage,
+    private core: CoreConnector,
+    private vpnDevice: VpnDevice) { }
+
+  public getSavedServers() {
+    const servers = this.loadServers().servers;
+    return Promise.all(Object.keys(servers).map((host) => {
+      return this.notifyCoreOfServer(servers[host].cloudTokens);
+    }));
   }
 
-  public addServer(code: AccessCode): Promise<Server> {
-    let token = parseInviteUrl(code);
-    if (!token) {
-      return Promise.reject(`Failed to parse access code: ${code}`);
+  private loadServers(): SavedServers {
+    return JSON.parse(this.storage.getItem(SERVERS_STORAGE_KEY)) || {
+      servers: {}
+    };
+  }
+
+  // Saves a server to storage, merging it with any already found there.
+  // Returns true if the server was not already in storage.
+  private saveServer(cloudTokens: cloud_social_provider.Invite) {
+    const savedServers = this.loadServers();
+    savedServers.servers[cloudTokens.host] = {
+      cloudTokens: cloudTokens
+    };
+    this.storage.setItem(SERVERS_STORAGE_KEY, JSON.stringify(savedServers));
+  }
+
+  public addServer(accessCode: AccessCode) {
+    // This is inspired by ui.ts but note that uProxy Air only
+    // supports v2 access codes which have just three fields:
+    //  - v
+    //  - networkName
+    //  - networkData
+    // TODO: accept only cloud access codes
+    const params: social.InviteTokenData = uparams(accessCode);
+    if (!(params || params.v ||
+      params.networkName || params.networkData)) {
+      return Promise.reject(new Error('could not decode URL'));
     }
-    // TODO: Do I need to wait for core.login()?
+
+    const cloudTokens: cloud_social_provider.Invite = JSON.parse(
+        jsurl.parse(<string>params.networkData));
+    this.saveServer(cloudTokens);
+    return this.notifyCoreOfServer(cloudTokens);
```

Comment:
Consider doing this on connection instead to reduce the scope where we need to worry about this core restriction.